rln-research/merkle-tree-update.md

```diff
@@ -0,0 +1,154 @@
+
+# Probelm Definition
+## On-chain tree storage
+In the current implementation of the rln, where the membership contract contains a Merkle tree holding the list of current members, the contract only supports the insertion of a new leaf, but not deletion. In addition to the lack of deletion support on the contract, it seems that even the most efficient scenario with logarithmic complexity would not be cost-efficient (since it involves almost 2.log(N) many expensive hashing which results in huge gas consumption.)
+
+
+**Hint**: While batch insertion can address the cost associated with insertion, the idea of batch deletion can not be applied since the deleted nodes are scattered across the tree (leaves).
+
+## Off-chain tree storage
+The discussion above brings us to the second possible approach in which the tree construction, storage, and maintenance will be delegated to the peers thus no tree gets stored on the chain. 
+The smart contract will only store an ordered list of existing members' public keys i.e., pk. In the list representation, when a member is deleted its pk is replaced with a special value. Peers need to listen to the events on the chain (i.e., insertion and deletion) to update their local constructed trees.
+
+In the off-chain tree storage, peers are responsible for the construction and maintenance of the membership tree (specifically the root). To do so, the peers must maintain enough number of nodes of the tree that enables them to calculate the updated tree root in the case of deletion and insertion of members. While an optimal solution exists to efficiently support insertion (the solution imposes only O(log(N)) storage complexity where n is the total number of members), efficiently supporting deletion is still an unknown. 
+
+One immediate solution is that the entire tree shall be stored by each peer, however, this solution is storage-inefficient due to the size of the tree. Provided that peers are resource-constrained, such a solution does not fit. Moreover, it will require peers to store the tree in a database and hence can negatively affect their performance.
+
+## Updating authentication paths
+In addition to the calculation of tree root, peers also need to update their authentication path based on the recent list of members i.e., the authentication path must be updated after each insertion and deletion. This is essential for peers' anonymity.
+An immediate solution for this is to store the entire tree on each peer, which, as we mentioned before, is not storage efficient and does not fit for the resource-constrained peers. 
+
+
+## Formal Problem Definition
```

Comment:
After chatting wtih Sanaz a bit we had a nice idea to improve here. 

The issue is that users need to consturct the merkle proof and path of all slashed and udpate nodes so taht they can change their own. So we had a nice idea to instead include the merkle proof as part of the slashing. This just adds a 32 (byte hash) * 32 (layers in merkle tree) = 1024 bytes which cost 3 gas per bytes(CALLDATALOAD) = 1024 * 3 = 3072 gas per slashing. Considering that we are already paying 20k gas per slashing this should not be too expensive to start out with. Later if it does get too expernvsive we can always update the nodes to track the tree and remove this reqruiement from data put on chain

This adds a few problems but they are easily solved 



1. Race conditions now that we have to incldue a merkle path in the latest merkle root two users getting slashed at teh same time will have merkle proofs for the same merkle root. This is fine because we can recontruct the full tree from just the oritignal proof. 
2. What if someone tries to slash a user but they include the wrong merkle proof, will this cause a fork. No this action fails and that users is not alshed.

Reply:
Thanks Barry for the comment!
I was discussing with Onur and realized one more potential issue  about calculating the auth path of the slashed member. If the slashed member does share her auth path, then everything will be fine, but  in case that the slashed member does not share her auth path willingly then the auth path should be calculated by some peer in an offline fashion. This will come with some computational cost when the slashed member has a high leaf index e.g., 2^31 (in a tree with depth 32), then computing her auth path involves a lot of hashings.
 A couple of thoughts about the solution:

1- We may consider an extra deposit at time of registration which will be withdrawn in case that the peer gets slashed and does **not** share her auth path
2- To query the auth path of a peer that is adjacent to the slashed peer (the sibling) in the membership tree (the slashed member and the adjacent peer will share the same auth path)
3- To query the system and fetch the tree nodes along the auth path (based on the nodes' GIDs). For each node of the tree at level j, there shall be 2^j many replicas (given that all peers always keep their auth paths updated) 
4- The tree may not get that much large in case groups are of smaller sizes, thus we need to deal with small size tree hence lower computational cost for slashing

Reply:
Could we check the merkle proof as part of slashing someone so no reward is paid unless the proof exists for any historic root.

Reply:
> What if someone tries to slash a user but they include the wrong merkle proof, will this cause a fork. No this action fails and that users is not alshed.

Assuming in that case, the root is __not__ calculated in the contract.

I think we can't rely on goodwill of a slasher or a member who can withdraw anytime they want.

> Could we check the merkle proof as part of slashing someone so no reward is paid unless the proof exists for any historic root?

Assuming in that case, the root is in the contract.

Single poseidon hash takes 21k gas and check a single proof will take more tahan 600k gas. So I think it will probably cost more than reward that slasher gets. Signups can be cheaper if we rely on batch registrations. I estimate that I would be around 40k gas which is actually not to far from cost of storing each pubkey. So I think we can also apply batch withdrawals (slashing) with another circuit and verify the new root. (btw I think we also can construct another circuit for larger batch registrations)


Another question is that should we consider optimistic approach for these operations?

Reply:
> Could we check the merkle proof as part of slashing someone so no reward is paid unless the proof exists for any historic root?

This would be against spam-protection, in specific, the peer (a spammer) who does double singling can escape the penalty by not providing its auth path hence preventing slashing.

Reply:
>  no reward is paid unless the proof exists for any historic root?

@barryWhiteHat
In this approach, who does check whether such proof exists? on-chain? or offline?

Reply:
> I think we can't rely on goodwill of a slasher or a member who can withdraw anytime they want.

Agree. But to pay the reward in a trustless way we have to 

1. Keep track of the merkle root
2. Merkle proofs at time of slashing. 

Otherwise we have no way to keep track of who is slashed. 

>  In this approach, who does check whether such proof exists? on-chain? or offline?

On chain. But as Onur says this is very expensive gas costs. We could use a snark to do this in a batch but this results in some race conditions around slashing that are annoying to solve.

Reply:
> I think optimistic approaches would leave room for other types of attacks (forks, inconsistency, etc.) and add complexity to protect against the subsequent issues. It may even result in more computational complexity than what we try to avoid (the cost of recalculating the auth path of the slashed member)

We could avoid some of these concerns by building on top of some EVM based optimistic rollup. Which would not really change anything we are building we would just deploy it on some optimistic rollup.